creduce/utils/testing.py

```diff
@@ -0,0 +1,847 @@
+            else:
+                logging.debug("delta test failure")
+
+                self._since_success += 1
+                self.pass_statistic.update(self._pass, success=False)
+
+                if (self.also_interesting is not None and
+                    test_env.check_result(self.also_interesting)):
+                    extra_dir = self._get_extra_dir("creduce_extra_", self.MAX_EXTRA_DIRS)
+
+                    if extra_dir is not None:
+                        shutil.move(test_env.path, extra_dir)
+                        logging.info("Created extra directory {} for you to look at later".format(extra_dir))
+
+            # Implicitly performs cleanup of temporary directories
+            test_env = None
+
+    def cleanup_results(self):
+        # Only keep unfinished environments or sucessful evironments with valid improvement size
+        self._environments = [env for env in self._environments if not env.has_result() or
+                                                                   (env.check_result(0) and
+                                                                    (self.max_improvement is None or
+                                                                     env.size_improvement <= self.max_improvement))]
+
+class ConservativeTestManager(AbstractTestManager):
+    pass
+
+class FastConservativeTestManager(ConservativeTestManager):
+    def process_results(self):
+        super().process_results()
+
+        # Seek for new speculative base test env
+        found = False
+
+        for i, test_env in enumerate(self._environments):
+            if (self.max_improvement is not None and
+                test_env.size_improvement > self.max_improvement):
+                continue
+
+            if test_env.check_result(0):
+                found = True
+                break
+
+        if found:
+            # Kill all later test envs as they are invalid anyway
+            # but keep this one in as it has to be processed "officially" in order
+            self.test_runner.kill(self._environments[(i + 1):])
+            self._environments[(i + 1):] = []
+
+            #FIXME: Need to move to create_env
+            self._base_test_env = test_env
+            shutil.copy(self._base_test_env.test_case_path, self._current_test_case)
+            self._base_test_env.state = self._pass.advance_on_success(test_env.test_case_path, self._base_test_env.state)
+            #logging.debug("Base state advance success: {}".format(self._base_test_env.state))
+
+class NonDeterministicTestManager(AbstractTestManager):
+    def can_create_test_env(self):
+        # Create new variants and launch tests as long as:
+        # (a) there has been no error and the transformation space is not exhausted,
+        if self._stopped:
+            return False
+
+        # (b) there are not already to many variants (FIXME: can potentionally be removed later),
+        if len(self._environments) > 200:
```

Comment:
Currently it is just an arbitrary number. I ran into the problem of too many open files since also the failed variants are kept until a successful one is found and everything is killed.
I think this problem could also occur in the Perl version but I will have to check the details.